### ISSUE TITLE
Resetting the whole library

### ISSUE THREAD

**Issue by lakeroe** [udoklein/dcf77#31]
Hello,

I'm running the library on an ARM STM32F103C8T6 and this works very well, but now I've noticed a strange behaviour.

After power-on the first received time if (clock_state >= Clock::free) is always valid and accurate but after some hours (or days) the time starts to drift away and sometimes also the date is wrong allthough the inaccuracy should be below 200 ms according to the comment in dcf77.h. This seems to happen especially in a quite noisy environment.

I assume this problem is related to: https://github.com/udoklein/dcf77/issues/27

Because I use this library to synchronize my ARM built-in realtime-clock I need a reliable time only once in a while. So my idea is to reset the library everytime a valid time is received (clock_state >= Clock::free).
My C++ knowledge is quite limited but I've managed to add a reset-function which calls DCF77_Clock::setup() and DCF77_Local_Clock::setup() and I call this reset-function everytime a valid time is received.

I'm currently running a test for about 24h and this seems to work well.

Do you think this is the right approach for my problem and is the reset-function correctly implemented ?

Thanks and best regards,
lakeroe

**Comment by udoklein** [comment]
My suspicion would be that this is caused by an unstable local oscillator. You are right that a similar issue was described https://github.com/udoklein/dcf77/issues/27 in the past. However I was never able to reproduce it at all. Without any means to reproduce it I have no way to locate the issue and fix it.

I suggest that you first try to tell the library that your oscillator might be unstable, e.g. set has_stable_ambient_temperature to false.

https://github.com/udoklein/dcf77/blob/1103b38718e8be451031f6b9986a4291c0a085e0/dcf77.h#L123

If this fixes the problem tell me about it. If this does not fix the problem I have no idea what is wrong. If you really need to enforce a reset of the libary this is in fact done by calling `DCF77_Clock::setup()`. This may work around your issue. So for you this might be good enough. For me this is really anoying because I still have no clue why my library fails for some people.

Most probably I need someone who can reproduce the issue and who lives close enough me such that I could visit and instrument the hardware with suitable measurement equipment.

**Comment by lakeroe** [comment]
Thanks for your reply.
I will do further tests and report back in a few days ...

**Comment by lakeroe** [comment]
I've done some more testing on two different hardwares:

Board 1: Arduino clone (but definitly quartz crystal based)
Board 2: STM32F103C8 blue pill board (http://wiki.stm32duino.com/index.php?title=Blue_Pill)

For all tests I used two different "Reichelt modules" and following time as a reference: https://www.uhrzeit.org/atomuhr.php

1st test: has_stable_ambient_temperature = false (as you suggested)
  I powered-up the boards and waited for the best possible signal quality (clock_state==synced). Then I wrapped the antennas in aluminium foil to disable dcf77-reception. I verified this by the non-blinking dcf77_monitor_led.
  Result after about 24h
  Board 1: clock deviation ~  0 seconds, correct date, clock_state=2 (free)
  Board 2: clock deviation ~ +4 seconds, correct date, clock_state=2 (free)

2nd test: has_stable_ambient_temperature = true
  Same procedure as before but changed software.
  Result after about 24h
  Board 1: clock deviation ~  0 seconds, correct date, clock_state=2 (free)
  Board 2: clock deviation ~ +4 seconds, correct date, clock_state=2 (free)

3rd test: has_stable_ambient_temperature = true
  I powered-up the boards and waited for the best possible signal quality (clock_state==synced). Then I put the antennas in an upright position to simulate worse dcf77-reception.
  Result after about 22h
  Board 1: clock deviation ~ +94 minutes, correct date,           clock_state=3 (unlocked)
  Board 2: clock deviation ~  -4 seconds, wrong date (14.7.2053), clock_state=4 (locked)
           exactly 62304s after power-on the date changes from 14.5.2018 to 14.7.2053

My modest opinion:
The first two tests show similar behaviour. Allthough the clock deviation on the second board is quite large it should be okay.
According to your github page a drift of up to 100 ppm is within the design margins of the library (4s in 24h is only about 46 ppm).
I think one problem is the library stays in clock_state 2 (free) even there's now reception for 24 hours.
One possible solution is to add a watchdog: If the input pin doesn't toggle for a specified amount of time it resets the library.

Test three shows the typical behaviour I mean. Board one shows large clock deviation and board 2 wrong date.
This test is not so meaningful as the result changes a bit everytime. For more reliable results more testing is needed.
But it should be quite easy to reproduce by simulating bad dcf77-reception (put the antenna near metal objects or partly wrap it in aluminium foil).

Maybe this way it's possible for you to reproduce the problem !

To finish with something different:
Further up you wrote resetting the library is done by calling DCF77_Clock::setup().
This does NOT work for me, I also have to call DCF77_Local_Clock::setup().

Thanks and best regards,
lakeroe

**Comment by udoklein** [comment]
Can you record some trace with the "Swiss Army Debug Helper" in mode Dm ? At least enough to reproduce the issue + some more hours?

With regards to the simulation of bad reception: I tried several things in the past. E.g. the standalone debug helper. So far I found no way to reliably simulate real word noise. I know that in theory a lot of the noise can be modelled, but reality is still different. Without real world data I have no chance. The only thing that I can say for sure that I can not reproduce the issue with my setup. And I tried quite a lot of things (e.g. putting the antenna on top of magnetic loudspeakers or on top of motors). Sometimes I just pulled the antenna for some hours. The issue is that this approach is very time consuming and I can not reproduce the behaviour so far.

**Comment by lakeroe** [comment]
Okay, I'll try to post a log to reproduce the behaviour.

Regarding to reset the whole library:
Does it make sense to you to also call DCF77_Local_Clock::setup() ?

**Comment by udoklein** [comment]
Good point. Yes, this is a good idea. Maybe I should add it as already suggested in issue #30. Well, maybe if it hits the third time ;)

**Comment by lakeroe** [comment]
Just a short intermediate result:
So far I was not able to reproduce the wrong date behaviour using the "Swiss Army Debug Helper".
But I'll keep trying ...

Furtermore there are two more questions:
1) In order to compile your library for an [ARM STM32F103C8T6](https://www.mikrocontroller.net/articles/STM32F103C8T6_STM32_Billig_Board) running at 72MHz  I added one
#if defined(__STM32F1__) #endif section at the end of [dcf77.cpp](https://github.com/udoklein/dcf77/files/2013164/dcf77.txt)
Could you just have a quick look to see if there's any obvious mistake ?
Is there anything to change for the different CPU frequency ?
2) In the meantime did you have a chance to look at my LOG above and do you have an explanation for the big clock deviation ?

**Comment by lakeroe** [comment]
Thanks for your analysis.
As I already wrote, I waited for a complete sync and then artifically introduced the signal fading by turning the antenna. My location is near [here](https://www.google.at/maps/@47.9358919,13.5932374).
For me this issue is not relevant at all, because I only need a reliable time once a day and this can be easily achieved by resetting the whole library as mentioned above.

Regarding to double check the local clock 100 ppm design limit. I don't have the possibility to accurate measure the quartz crystal frequency but how about following idea:
I use the 1000 Hz Systick Timer to create a clock and compare it to a reference (www.uhrzeit.org/atomuhr.php). If it deviates less than 8,6s in 24h (=100 ppm) then I assume it's okay ?

**Comment by nameoftherose** [comment]
We are just a few kilometers apart ...
What dcf module are you using? How long does it take to sync?
How long it stays in sync per day? (in my case about 8h/d)

**Comment by udoklein** [comment]
Just send me your phone number by private mail. Maybe we can have a call during the weekend. With regard to the issue, you say you use systick. Thus I assume you are running on ARM. On Saturday I will setup a test with an ARM board for 24h and see if there are any regressions. With my board I had never any issues so far as the signal quality around here is way above the limit for my library. So either there is some regression that did not catch my attention or there is something with your setup. 100 ppm should be absolutely OK, but there may be other issues. I suggest to discuss this by phone (or in person).

**Comment by udoklein** [comment]
So far my tests show nothing unusual. Maybe you could run the debug helper in mode DA to see if your clock is really stable. Sometimes it is not the hardware but interrupts that mess up the timing.

**Comment by udoklein** [comment]
After 48 hours on Arduino Due (ARM) with a Pollin Module everything looks perfectly well.

``   168310, +------XXXXXXXXXXXXXXXXXXXX85-+---------+---------+2X9--7---+---------+3--------+---------X---------
confirmed_precision ?? adjustment, deviation, elapsed
0.0625 ppm, @+ , 24.8750 ppm, -1 ticks, 412 min + 38007 ticks mod 60000
57/100~25/600

Decoded time: 18-05-27 7 21:31:43 CEST ..
  18.05.27(7,0)21:32:42 MEZ 0,0 26 p(67457-40:255) s(191-0:26) m(244-5:33) h(241-0:33) wd(249-1:34) D(246-3:34) M(245-1:34) Y(250-4:34) 127,32,32,50
``

I have to assume that it is something with your setup. The question is what?

**Comment by lakeroe** [comment]
Sorry for my late reply. I was not at home the last days ...

I could observe the drifting clock on both AVR and ARM based board.
One thing I noticed is to have a clean and stable power supply. When powering the whole setup by a very cheap "usb china power supply" I don't get any reception at all.

As suggested, I will run the debug helper in mode DA and report back.
Furthermore I can offer to send you a complete ARM based affected hardware for your investigations.

I'm quite busy until sunday but next week we can have a call if you like.

@nameoftherose
If you're interested you can send me your exact location by private mail ...

**Comment by lakeroe** [comment]
Please find another LOG attached (in DA mode).
Interestingly the date is wrong but the time is accurate.

[dcf_2018-05-28_ARM.zip](https://github.com/udoklein/dcf77/files/2049939/dcf_2018-05-28_ARM.zip)

**Comment by nameoftherose** [comment]
This looks very much like with what I experienced. It is not clock drifting, the decoders decode the wrong date.  If computed day and decoded day are not equal the date decoders should be reset and transition to synced state barred. I also experienced hour errors, but very rarely.

**Comment by nameoftherose** [comment]
This is a graph of the various performance metrics produced by swiss_army_debug_helper.
![qf](https://user-images.githubusercontent.com/18288708/40713925-e147228a-6409-11e8-9a18-41eaab3f2b13.jpg)
The phase_lock (marked p) is erratic. Assuming the library has been properly ported, I would say it is noise. Filter power supply - I am using LC filters - even use battery or feed from a laptop operating from its battery. Reset the clock whenever calculated week_day defers from the decoded one.
For my location (2000km from the transmitter) a 55mm antenna is too short, but you are much closer.

**Comment by lakeroe** [comment]
Thanks for your analysis, you can clearly see the dependancy of the antenna orientation.
At the beginning the phase is locked for some time (antenna is horizontal and pointing towards dcf77 sender).
Then it looses the lock (antenna was turned vertical) and at the end it's locked again (because antenna is horizontal and pointing towards dcf77 sender again).

**Comment by nameoftherose** [comment]
I am using the [Conrad module](https://www.conrad.de/de/dcf-empfaenger-modul-641138-passend-fuer-serie-c-control-641138.html). Its antenna is 50mm.

**Comment by udoklein** [comment]
Looking at the analysis of nameoftherose there is very interesting behaviour. First sof all the phase lock fluctuates a lot. Then during the period with the poorest phase the month and day decoder "lock" to the noise. Once the signal gets really good (before 17:00) the other decoders increase in quality. However the mistakenly locked decoders start to converge to the proper value. Hence the quality decreases.

The easiest way to fix this mess is to not allow the "slower" decoders to decode before the fast decoders are ready. That is: require that locks can only be acquired in the order phase, seconds, minutes, hours, days, months, years.

I think this would fix it. There is a catch though. The current approach allows for significantly faster startup. I think I will fix it by introducing another configuration flag. What do you think about this proposal?

With regards to using the weekday as some kind of checksum this will unfortunately not fix it. The reason is that if your signal quality is such that this issue happens then introducing this as a checksum will fix it for 6 out of 7 days of the week. The rest there will be still this issue.

There is one more thing that I could do: I could introduce additional "flat out detection bins". The price to pay is 1-2 additional bytes of sram consumption per decoder. Thus about 12 additional bytes. This would also help to relieve this issues.

Thus the action plan would be as follows:

1) Add a flag "optimistic decoder" which toggles the way the decoders are working. TRUE = as currently implemented, FALSE = do not allow the slow decoders to decode before the fast ones are at least marginally ready.

2) If the flag is TRUE use the weekday information for faster transition to "synced", otherwise require improved consistency.

3) Add a flag "flat out detector". If FALSE go ahead as today. If TRUE add additional detection bins for flat out 0 and flat out 1. (but not for minutes or hours, for those only for flat out 1). If the decoder locks to the flat out signal then void all successive decoder stages and do not allow to sync.

What do you think about this approach? In particular about the choice of the default values? Should I default to maximum robustness (paying with more memory and slower initial sync) or should I default to the optimistic setup and require that people with extra poor reception need to adapt the configuration?

**Comment by udoklein** [comment]
@nameoftherose By the way: what did you use to plot the log statistics?

**Comment by udoklein** [comment]
@lakeroe If you send me your hardware then I also might be able to gain more insights. Did I understand it right that you are within <300 km from Frankfurt?

One more question: why are you putting the antenna intentionally into a poor reception position? This is a very interesting approach but I would assume that the antenna is mounted in an optimal orientation and that the library will only have to tackle "other noise". What you are doing is basically raising the noise floor by more than 10d dB, maybe even 20 dB. This is great for testing but close to Frankfurt it is somewhat pointless. So I am wondering why you do this?

BTW: together with the analysis of nameoftherose I think this finally shed some light on why some people reported issues with my library that I was not able to reproduce. Obviously your test approach is a good idea :)

**Comment by udoklein** [comment]
I analyzed the log and with the hints of the picture by nameoftherose the plot thinkens. Your decoder module is biased to all 1. Thus during periods of bad reception it shows a different behaviour than mine. Mine biases to 0. This also explains why I was never able to figure this out. My module is biased differently. I will fix the library. However I am very busy right now. I can not promise that I will be finished in June. This might take some weeks.

**Comment by nameoftherose** [comment]
The graph is created as follows:
1. minicom creates a timestamped log
2. the log is proccessed with awk -f [qf.awk](https://github.com/nameoftherose/dcf77-leonardo_support/blob/master-1/logfiles/qf.awk) to produce qf.csv
3. qf.csv is opened in gnumeric
4. [qf.gnumeric](https://github.com/nameoftherose/dcf77-leonardo_support/blob/master-1/logfiles/qf.gnumeric) is then opened, it contains links to qf.csv, and the graph is shown
5. the dates in cells A1 A2 are the graph limits.
This could be done in any spreadsheet.

**Comment by nameoftherose** [comment]
Yes I think your action plan is very promising.
Take your time.
Thank you.

**Comment by lakeroe** [comment]
@udoklein
Regarding your questions

*) I think your plan sounds promising and we should give it a try.
Once the new software is available for sure I can run further tests and post logs.

*) Default values
I would go for maximum robustness because I think it's more important to have a reliable time than to save some bytes on RAM and faster sync.

*) My hardware
Due to the newest findings do you think it's still necessary to send my hardware to you ?
If yes, another possibility is to by your own hardware. It's just so cheap and probably the same price than sending mine to you and back. The hardware I use is
[Pollin modul](https://www.pollin.at/p/dcf-empfangsmodul-dcf1-810054) 5,45€ + 4,95€ (shipping)
[STM32F103C8 blue pill board](https://de.aliexpress.com/item/STM32F103C8T6-Micro-USB-controller-STM32-Development-ARM-Learning-Board/32832747542.html) 1,85€ from china
Is that okay for you ?

*) Distance from Frankfurt
According to Google Maps I'm 407km away (in Austria near Salzburg).

*) Antenna orientation
I put the antenna in a poor reception position to get a quicker result.
If it's in an optimal position I might never reproduce the drifting time problem ...

*) Time schedule
No hurry and take your time. It's finished when it's finished !

**Comment by udoklein** [comment]
Well, I used a Pollin module for my tests --> in my location it is differently biased. As you say most probably I will gain no insights from your hardware. However you could do me a favour. Now that we know what is most likely the issue it would be nice if we can create a log file which will reproduce the issue with my standalone debug helper. Unfortunately the log file with the issue does not. The reason is that the log does not capture the full information prior to the first sync.

There would be two options:
1) I synthesize this information
2) Hopefully easier: you create another log just the way you did when you reproduced the error. But this time instead of turning the antenna after the first sync give it another 20-30 minutes to acquire a better lock, then turn the antenna and wait till the issue reoccurs.

According to the theory (2) should still reproduce the error but it should also capture enough signal such that the standalone debug helper will also recreate the issue. This would help me a lot during testing.

Would this be possible for you? I am not in a hurry. If this takes 2 or 3 weeks it would still be in time for testing.

**Comment by lakeroe** [comment]
Sure, I will create another log and report back ...

**Comment by lakeroe** [comment]
Please find another LOG attached (in Dm mode).
This time both date and time was wrong ...
Maybe @nameoftherose is kind enough to generate another nice looking graph ...

@udoklein
Could you just have a quick look if it fits your need ?

[dcf_2018-06-05_ARM.zip](https://github.com/udoklein/dcf77/files/2077726/dcf_2018-06-05_ARM.zip)

**Comment by nameoftherose** [comment]
![qf](https://user-images.githubusercontent.com/18288708/41061223-8a0c65da-69da-11e8-9ded-3651db87beab.jpg)
I  would not say the clock is drifting, the algorithm decodes the wrong time (line 638370) and syncs to it. Synchronization is then lost and reacquired, this time correctly (line 658922).
This log confirms @udoklein analysis.

**Comment by udoklein** [comment]
@lakeroe - Excellent. The file is does reproduce the issue with the standalone debug helper. As it now turns out even the first one did but I mistakenly used a wrong parameter setup. Sorry for causing duplicate work. However on the upside I have now more test data :)

With regards to the plots I added a log_2_csv converter in the latest commit here: https://github.com/udoklein/dcf77/blob/master/extras/standalone_debug_helper/log_2_csv.py  
The tool has a command can convert logs to csv files. Then any plot program should be able to consume it. Thanks to @nameoftherose for bringing up this idea. I still feel somewhat humbled that this did not occur to me earlier.

**Attention:** I also changed the log format slightly. Thus it does not work with version of the library older than v3.2.8. In case of logs written with previous versions of the library all occurences of "MEZ" must be replaced by "CET" and all occurences of "MESZ" must be replaced by "CEST" before converting log files.

**Comment by udoklein** [comment]
By now I think I will patch 

`static void process_single_tick_data(const DCF77::tick_t tick_data)`

The point is that in

`static void sync_lost_event_handler()`

I wipe the slower decoders once a sync is lost. But then I do not prevent them from gaining "quality" even if the fast ones fail to do so. The point is that this situation is with exceedingly high probability caused by biased noise. If the noise is random, then this can not really happen.

As a consequence I would not need an additional flat out detector at all. The fast stages already provide this information.

I am now pondering what happens if the fast stages have very high quality and the slow stages are not yet there. Now if biased noise enters the clock then I can not detect it and it will again pull the slow stages into this undesired state.

Of course I could add a check for "jumps" in time. But then what to do? Maybe reset the clock.

I am still thinking. The implementation of the fix will be much easier than finding it.

**Comment by udoklein** [comment]
On the other hand maybe a flat out detector could leverage the timezone bits + the 1 minute digit bit. Thus every two minutes it should see 3 ones and 3 zeroes. If this is not the case then the signal might be biased. Now I need to figure out how to prevent it from a "random walk". Probably an exponential filter approach would be good. This raises the question for a suitable filter constant.

**Comment by udoklein** [comment]
With regards to the flat out detector it might be actually better to not look into any of the decoded bits. Probably it would be best to look into the raw signal. E.g. it might be a good idea to look at the first 100 ms and the 500-600 ms interval. The pattern might be

0 0 --> flat 0
1 1 --> flat 1
0 1 --> drowned by noise
1 0 --> everything the way it should be

This would have the significant advantage of a very short latency for detecting flat outs. This leaves the question what would be a good fitler constant. --> Need to analyze the data of the examples.

**Comment by Jo-Achim** [comment]
Great work here!

**Comment by udoklein** [comment]
Release [v3.2.9](https://github.com/udoklein/dcf77/tree/v3.2.9) introduces some slight change in the way the decoder templates are parametrized. This "optimization" is part of the preparation for the fix of the current issue.

**Comment by udoklein** [comment]
I am right now testing v3.2.10. Here are some results if I feed the logs lakeroe into it.

![01_dcf_2018-05-15_avr_v_3_2_10](https://user-images.githubusercontent.com/5356932/42124034-4ac9f01a-7c5c-11e8-9e9d-89809969d0d7.png)
![02_dcf_2018-05-18_arm_v3_2_10](https://user-images.githubusercontent.com/5356932/42124038-5f5ed432-7c5c-11e8-9e6e-8f2a08f60af3.png)
![03_dcf_2018-06-05_arm_v3_2_8](https://user-images.githubusercontent.com/5356932/42124040-626c0320-7c5c-11e8-8270-62155428d3a7.png)
![03_dcf_2018-06-05_arm_v3_2_10](https://user-images.githubusercontent.com/5356932/42124041-64384e02-7c5c-11e8-9eab-2c2196a20407.png)
![03_dcf_2018-06-05_arm_v3_2_10_seconds](https://user-images.githubusercontent.com/5356932/42124042-65e3bc3c-7c5c-11e8-87f4-97fbc770a7ed.png)

**Comment by udoklein** [comment]
Picture 1 and 2 show the new library for the olders two files. Picture 3 and 4 show library 3.2.8 vs. 3.2.10 vs. the newest log file. As it turns out the library still struggles with this file. Picture 5 shows why. I took only every 60th sample. Thus the second should be stable. However it is not. The library acquired a phase lock but the second is drifting. What ???????

I double checked the original log file. This is not a bug in the library. This is the hardware which is causing the issue. Actually a one point between 86000 and 88000 seconds its drifts by almost 1 second in 500 seconds or 2000 ppm. Way out of spec.

I would like to point out two things:
1) In the last case there is either some hardware flaw with the local oscillator or there is some other code running that messes with the interrupts.
2) Because of the fix the new library takes longer to actually lock to a signal. Thus if there are only very short periods with a reasonable good signal, then it may not lock to the signal at all. There is not a lot that I can do about this. The library is statistical in nature. Everything that I do to reject bad signals will decrease the chances to accept correct but noisy signals. I hope the change is an overall improvement.

Please test and provide feedback.

**Comment by udoklein** [comment]
If you are testing either use [v3.2.10](https://github.com/udoklein/dcf77/releases/tag/v3.2.10) or, if you use v3.2.11 make sure that you set https://github.com/udoklein/dcf77/blob/v3.2.11/dcf77.h#L110
to the value used in v3.2.10. E.g.

`enum controller_minute_quality_threshold_t : uint8_t { aggressive_minute_quality = 0, standard_minute_quality = 2, conservative_minute_quality = 4, paranoid_minute_quality = 6 };
    static const uint8_t unacceptable_minute_decoder_quality = controller_minute_quality_threshold_t::conservative_minute_quality;
`

**Comment by brettoliver** [comment]
Hi Udo.
Thanks for the update. 
I am in the south of England. This time of year once the temperature rises I get very bad signal reception in the afternoon into the evening.  
I have 9 DCF77 clocks running off 1 Aerial via a Digital repeater so they all get the same signal. One clock runs library version 2 and never has any issues the other clocks run the old version 3 library. One of the clocks has a problem when the signal gets bad over a long period and losses Sync showing Lock on the display. Eventually the time and date go way out and the clock never returns to sync even with a good signal.

I have now loaded this clock with you new library 3.2.10 and will see how it goes. 
The DCF77 Signal is very bad at the moment so I expect a long wait for sync but I don't mind this if it fixes the Sync issues.

I'll let you know how I get on.


Regards.
Brett.

**Comment by udoklein** [comment]
Hi Brett,
if you could record one or two days with "Swiss Army Debug Helper" in mode Dm, then I could analyze what happens. This is the only way for me to get real world noise. As it becomes more and more obvious my simulator does not simulate the real world situation. Hence I require these logs to improve the library further.
Regards, Udo
P.S. Which exact version 2 are you using? Maybe I can spot the difference in the code and port the different behaviour to version 3 again.

**Comment by brettoliver** [comment]
Hi Udo.
I will be moving the clock that has sync issues in the next couple of weeks. I will set it up near my PC and record the data for you.
The version 2 code I use on my Master Clock is v2.0.4. I think this has been running for over 4 years now with no errors.
Regards.
Brett

**Comment by brettoliver** [comment]
Hi Udo.
My clock has now synchronised after 8 hours of intermittent bad signal. See my signal here 1 = 100% 0 is <100% http://home.btconnect.com/brettoliver1/Geiger_Counter/Charts.htm

The only problem is the clock has synchronised to the wrong time. The seconds are correct but the time shows  06:09:45 20/04/2000 not 20:09:45 30/06/2018.

I would normally run an EEPROM erase program at this stage to get the clock to sync correctly. 
 http://home.btconnect.com/brettoliver1/Arduino_EEPROM_Erase.htm

What do you think?
Brett.

**Comment by brettoliver** [comment]
Have run EEPROM Erase as clock went to Lock. Clock has now  syn'd correctly after only a few minutes.

**Comment by udoklein** [comment]
Hi Brett,
I have no idea why erasing the EEPROM would help. Maybe this is just coincidence, maybe not. I definitely need log files to gain more insight. I pushed a new version of the library (v3.2.12) with slightly enhanced logging. If you could use this and create some log files maybe I can gain more insights.
Regards,
Udo

**Comment by brettoliver** [comment]
Hi Udo.
I nearly always have to erase the the EEPROM once any of my clocks have lost sync. I presumed the learned "settings" were stored in EEPROM and got reloaded when the Arduino was restarted. Once these "settings" are incorrect I find my clocks won't sync to the correct time even after reset.
Once I do an EEPROM erase the clocks always re-sync correctly.

I will be able to use your "Swiss Army Debug Helper" now as I have found a long USB cable to reach my clock (it's a Longcase Clock so I can't move it easily).

I have loaded the  "Swiss Army Debug Helper" into my clock and sent "Dm" to the clock.
I am showing a "scope" type output on the serial monitor.

How do I save a log file?
Brett.

**Comment by lakeroe** [comment]
Hi Udo,

I tried v3.2.11 (with default parameters) and ran my "usual" test (put the antenna in an optimal position, wait for the best clock_state possible and then put the antenna in a non-optimal position). After 33229 seconds the library decodes the wrong time ...

Could this be due to local clock drift you assumed above ?

Best Regards,
lakeroe

LOG-File: [dcf_2018-07-02_ARM.zip](https://github.com/udoklein/dcf77/files/2154009/dcf_2018-07-02_ARM.zip)

**Comment by lakeroe** [comment]
In my first understanding this setting only effects the initial lock time, so I've left it at the default value.
But now I think you're right.
So I change this setting and do some more testing ...

**Comment by Jo-Achim** [comment]
Hello Udo,
hello together,

I would like to get involved in this topic, although I did not understand all the details of the problem / solution.

In my understanding, the current problem solution (v3.2.1x) has, compared to the 'normal'-decoder-mode, the problem of a very long initial decoding of the DCF signal, which is, why the maintenance of the 'normal' configuration of the decoder (Out-of-the Box) makes sense.
Only in especially bad reception situations with corresponding noise is a configuration change of the decoder to the new configuration (v.3.2.1x) useful. Correct?

So my question: Can the decoder be changed during operation to the new configuration; e.g. about changing a variable?
If so, would be the following solution possible? ...
The decoder starts in 'normal' mode and stays in it (the first decoding always seems to be correct!?).
It is up to the user whether and how he integrates a kind of time and date plausibility check into his sketch. For a wrong year, it is probably the easiest; everything that is not a leap from year + 1 and on 31.12., is wrong - and leads to switching the decoder in the 'safest mode', i.e. the current solution (v 3.2.12).

The next question is, whether there must be a switch back to the 'normal' mode of the Decoder? Which raises the question of whether a solution might also consist in the fact, that the decoder after first decoding switches in the 'safest mode' and remains there?
If this is not ideal or desired, the Date / Time plausibility test must downshift in the 'normal' mode within appropriate intervals.

If it should be possible that the decoder could independently determine a signal quality that leads to the date / time error - which I do not expect because of the effort - then the decoder could ideally perform the toggling itself.

Maybe there will be more approaches...

**Comment by udoklein** [comment]
@Jo-Achim as you stated yourself you lack understanding of the issue. We have no lack of questions or ideas. If you want to contribute, please tell me:

- What is your background? 
  - Do you have C++ coding skills? Which level?
  - Do you have signal processing, statistical or mathmatical skills? Which level?
  - What else do you think you could contribute? 
- How much effort would you be willing to contribute?
- What is your stake in this project? Why do you want to contribute?

**Comment by lakeroe** [comment]
I did some more testing by setting the unacceptable_minute_decoder_quality variable to different values.
Furthermore I used 3 different antenna positions:

1. Best position  - antenna horizontal and rotated by 90 degrees towards Frankfurt
2. Worse position - antenna vertical
3. Worst position - antenna horizontal and pointing directly towards Frankfurt

Results

1) aggressive_minute_quality
Time for initial lock (correctly decoded date/time) in antenna position 1: 546 s
Antenna position 2: Wrong date/time after 33229 s (old behaviour)

LOG-File: [dcf_2018-07-01_ARM.zip](https://github.com/udoklein/dcf77/files/2165264/dcf_2018-07-01_ARM.zip)

2) standard_minute_quality
Time for initial lock (correctly decoded date/time) in antenna position 1: 986 s
Antenna position 2: No time drift
Antenna position 3: time starts to drift after a few minutes

LOG-File: [dcf_2018-07-02_ARM.zip](https://github.com/udoklein/dcf77/files/2165265/dcf_2018-07-02_ARM.zip)

3) conservative_minute_quality
Time for initial lock (correctly decoded date/time) in antenna position 1: 1618 s
Antenna position 3: time starts to drift away (off by about 130 s at the end of the log)

LOG-File: [dcf_2018-07-03_ARM.zip](https://github.com/udoklein/dcf77/files/2165266/dcf_2018-07-03_ARM.zip)

4) paranoid_minute_quality
Time for initial lock (correctly decoded date/time) in antenna position 1: 1720 s
Antenna position 3: time starts to drift away (off by about 130 s at the end of the log)

LOG-File: [dcf_2018-07-04_ARM.zip](https://github.com/udoklein/dcf77/files/2165263/dcf_2018-07-04_ARM.zip)

So in my opinion you can clearly see an improvement although the time drift is still present.
But I still miss one feature:
Resetting the whole library (see the title of this thread) is still not implemented ;-(

However a big thanks for your great work here.

**Comment by nameoftherose** [comment]
@lakeroe , is position 3 a new position or have you used it in the past?
What antenna position was used in the previous logs?
In log no 2 how long have you kept the antenna in position 2?
A drift of 130s in ~8.5hours corresponds to 4300ppm. This is way too much. Are you sure your PC remained synchronized during the measurement period?

**Comment by lakeroe** [comment]
> Is position 3 a new position or have you used it in the past?

Position 3 is not new, I've also used it in the past sometimes.
However during those tests I found this is even worse than putting the antenna in an upright position.

> What antenna position was used in the previous logs?

I used position 2 (at least I think so).

> In log no 2 how long have you kept the antenna in position 2?

I think about 1h in position 1, 23h in position 2 and the rest in position 3
In logs no 3 and 4 I skipped position 2 completely.

> A drift of 130s in ~8.5hours corresponds to 4300ppm. This is way too much. Are you sure your PC remained synchronized during the measurement period?

Yes I'm sure, my reference time is this: https://www.uhrzeit.org/atomuhr.php

**Comment by nameoftherose** [comment]
@lakeroe 

>However during those tests I found this is even worse than putting the antenna in an upright  position.

 It seems to me you are changing the rules. For position 2, which was the standard so far, this is a huge improvement. Have you encountered in practice a situation worse than position 2?
Bear in mind that due to the statistical nature of the problem no algorithm will be perfect. We can only lower the probability of error.
The behaviour in logs 3,4 is very different from what I observe, therefore I cannot comment.

**Comment by brettoliver** [comment]
Sorry everyone, I seem to have lost some notifications on this thread and missed some replies.

Udo said 3 days ago.
"@brettoliver: the behaviour you are experiencing might be a combination of two issues:

You local oscillator is somewhat unstable. In particular it might have a significantly different frequency after a reset vs. long runtine.
Due to the autotuning with the large tolerance it tunes to the frequency that settles after a long runtime. Then after a reset it will try to immediately lock to the persisted frequency. This is of course impossible if the oscillator drifted a lot.
If my theory is right then this can be remedied as follows:
In dcf77.h change
static const bool has_stable_ambient_temperature = true;
to
static const bool has_stable_ambient_temperature = false;
and tell me if this fixes the issue. If not so I will change the library such that for "false" it will not persist the tuned frequency anymore. This is the "big hammer" option that I would like to avoid but this effectively resets the EEPROM on each new start."


Udo.
I have reconnected my Longcase Clock DCF decoder board and loaded the code back with your modification to the dcf77.h file.

I will run it over the next few days and see how it goes.

Your theory sounds good. I always wondered why some of my nine DCF connected clocks although connected to an identical signal (I use a digital repeater off 1 common DCF77 receiver) would fail and not re-sync after a long period of bad signal.

I like the "big hammer" option as a reset will always bring a clock back to life without having to run the EEPROM clear program. This is especially true if many years down the line you have an old clock running outdated library versions fails and you have no compatible hardware/software to perform an EEPROM clear and reload the Arduino.

**Comment by udoklein** [comment]
@brettoliver [v3.3.0](https://github.com/udoklein/dcf77/tree/v3.3.0) introduces the biggest possible hammer for your issue. I completely removed the EEPROM persistency.

@lakeroe thanks for the logs. I may need some time till I can analyze them. I will be somewhat busy the next several weeks. @nameoftherose is most probably right though. And if your oscillator drifts significantly more than 100 ppm, then you are using the library way out of spec. As a consequence there are no guarantees whatsoever. I would expect that it only decoded garbage. @nameoftherose is also right about the statistical nature of the library. There is no simple means for me to sort out the proper signal if the quality is poor. If the target platform would be a bigger system with lots of CPU and memory and not some poor microcontroller with 2k SRAM, then this would be much easier. However the whole idea was to provide a very noise tolerant library in spite of such constrained resources. 

If the logs indicate that my fix was not really good, I will try to improve it. But I doubt that I can much more. 

Conclusion: optimize your antenna position.

**Comment by nameoftherose** [comment]
@lakeroe , Have you observed drifting with ATMEGA? Will you consider creating a timestamped log? (but only if you are **absolutely sure** the logging computer maintains **correct** time)

**Comment by udoklein** [comment]
@lakeroe @nameoftherose In my opinion the drift is clearly visible in the logs. No need for a timestamped log. The point is that the scope view shows that it drifts. As long as the library maintains its phase lock it is clear that the signal is "timestamped" by DCF77. Thus by definition (because DCF77 transmits the legal time as maintained by the PTB)  any shift of the phase is a deviation from nominal frequency.

**Comment by udoklein** [comment]
Well, this is the point. The library was ported by me (and tested) for Arduino Due. I had it under test several weeks. So I assume it should be OK. However lakeroe seems to be using a different board. On the other side he also tested with AVR.

Once I find a little bit more time I will analyze the new logs and see what I can find.

**Comment by Jo-Achim** [comment]
One, despite my ignorance ;-), maybe helpful random observation as information:

Unknown, therefore hardly repeatable and so first occurred circumstances, led to the fact that two quartz MiniPro (each ~ 30ppm) with the same software but different DCF77-Libraries different times and a different date indicated. Both MiniPros are powered by the same DCF77 receiver!

**MiniPro_1 with Library <= v. 3.2.8 and >= 3.2.2** (out-of-the-box), unchanged for some time (therefore rather < v. 3.2.8), shows - verified - the correct time (and date).
**MiniPro_2 with Library v. 3.3.0** (out-of-the-box) initially shows only an incorrect time, later also an incorrect date; EEPROM explicitly deleted.

What happened? An inadequate, very noisy DCF77 signal resulted in 'unlocked' reception conditions within about 24 hours.

MiniPro_1 was initially affected; Synced -> Locked -> Unlocked. MiniPro_2 followed clearly delayed; Synced -> Locked -> Unlocked. Worst decoder mode for both: Unlocked.
The time interval with the MiniPro_2 followed is unknown (the decoder modes are signaled by colored LEDs, time or date were not checked at first). But it is from 10 minutes and (much) more likely to go out; it could also have been 1 - 2 or more hours.

After returning to perfect reception conditions, MiniPro_1 was the first to return to sync mode; an incorrect time or date was not observed.
While MiniPro_1 was back in sync mode, MiniPro_2 was still in unlocked mode. 

In the individual observations consistently good reception conditions, the following data emerged (the date of both MiniPro was on Wed., 11.07.2018 identical correct):

**Wed., 11.07.2018:**
**MiniPro_1 = 12:43:10 (synced);	MiniPro_2 = 11:58:36, Wed., 11.07.18 (unlocked).** 
MiniPro_1 = 13:00:00 (synced); 	MiniPro_2 = 12:15:26, Wed., 11.07.18 (unlocked).
  ...
MiniPro_1 = 13:45:00 (synced); 	MiniPro_2 = 13:00:26, Wed., 11.07.18 (unlocked).

MiniPro_1 = 13:51:00 (synced); 	**MiniPro_2 = 13:06:26, Wed., 11.07.18 (locked).**
MiniPro_1 = 14:00:00 (synced); 	MiniPro_2 = 13:15:26, Wed., 11.07.18 (locked). 
  ...
MiniPro_1 = 21:00:00 (synced); 	MiniPro_2 = 20:15:26, Wed., 11.07.18 (locked).

<no further observations on 11.07.2018>


**Thu., 12.07.2018** (time and date of MiniPro_1 still correct):
MiniPro_1 = 03:29:04 (synced); 	**MiniPro_2 = 03:04:05, Tue., 18.07.79 (locked).** 
MiniPro_1 = 08:02:07 (synced); 	MiniPro_2 = 07:37:08, Tue., 18.07.79 (locked). 
  ... 
MiniPro_1 = 14:00:00 (synced); 	MiniPro_2 = 13:35:01, Tue., 18.07.79 (locked). 
  ...

The above last state continues to persist; including time and date offset. I will supplement if MiniPro_2 has reached 'synced' mode to estimate the total time until then.

Notes: The synchronization bit is error-free "0" (without visible interference). Initial initialisation takes about 6 - 7 minutes with approximately the same good reception conditions.

Conclusion: According to my impression, the Library v. 3.3.0 (compared to previous versions) buys the longer stability with deteriorating signal with a very long time back into sync mode; with - in this case - wrong time and date.

And sorry, no logs.

**Comment by udoklein** [comment]
@Jo-Achim Your "analysis" is not helpful. Either you provide logs for analysis or you shut up. I asked some pretty simple questions about your capabilities that you did not answer so far. Please provide some answers to my questions or provide reasonable input for analysis. Please do not provide insights like "the improved version of the library is somewhat better but not good enough". This does not lead to any insights and does not help with any analysis at all. If you continue "contribute" the way you do right now I will remove you from this conversation.

**Comment by Jo-Achim** [comment]
@udoklein : Ok, Udo, ignore my posts - and if you think its necessary, delete them.

But instead of explaining to us, why my post - in line with the reported problems - was "not helpful", you ask about my posts only after my qualification, my findings, etc !?
But I can not see a reason that underpins "unhelpful" or deficiencies in my "information". I have just communicated a finding - completely without scientific claim.
So if that's your way of dealing with experiences, then please.
And if you do not even come up with the idea that an earlier version of your library might be better than the latest ones because of the posted data, then I actually do not understand something.

Your own statement about version 3.3.0 was: "The frequency tuning mechanism was too ambitious. **This caused more issues than it solved.**" Where is the problem? Stressed, little time? - Everything OK!

Finally, there was a behavioral comparison between two of your libraries. Of course, I apologize for not being able to recognize that the situation described would occur, so I did not log the whole thing, etc. But it was, therefore I apologized for the missing logs, an unpredictable situation.
Notwithstanding that not everyone (user) has the time and equipment to log days of logs, and so on.

And that's also clear: your DCF77 libraries are the best I've found. But that does not mean they are flawless despite their level.
As a developer, I think you should be pleased with questions / criticisms, as it is impossible within certain limits to invent / define and test all "framework conditions" - and ultimately improve the quality of your own work. Even if something in the test turns out to be "irrelevant".

At the moment, the impression remains that you can hardly handle criticism of people below your level (as a developer of your library). You certainly remember a variety of email contacts in the past that already addressed the problem of this thread (wrong date and time) - but without (mutually) to gain sufficient knowledge or to be able to provide.

In sum, there is no reason in my opinion (for you) to feel personally attacked. But critical questions, which could possibly be completely unfounded, must be possible. Just like 'randomly generated information / insights' would be worthy of professional judgment / evaluation.
The questioning of the qualification of the informant is justified only if its information allows a verifiable suspicion of defectiveness.

In this sense. (Take a break, take a deep breath, take your time, ... might be helpful.)

**Comment by udoklein** [comment]
@Jo-Achim In Germany we have a saying "Das Gegenteil von gut gemacht ist gut gemeint.". This translates to something like "The converse of well done is well intentioned". Your help is well intentioned but actually not well done. You might want to read http://www.catb.org/~esr/faqs/smart-questions.html why. You do not provide any actionable insights but are wasting bandwidth. If you find an issue with any of the libraries, nail it down and put it into the issue tracker. If you can not nail it down then provide logs. If you want to get personal forget it.

**Comment by Jo-Achim** [comment]
@Udo: Well, I had already spoken of "Unknown, therefore hardly repeatable and so first occurred circumstances" and therefore regretted to have no logs.

Since the same library - only in different versions but identical environment (the same DCF77 receiver supplies both (identical) MiniPros in parallel) - behaved significantly differently, this could have been a clue, regardless of my abilities. In this respect, a mistake or a deterioration of the newer library would not be completely excluded.

I do not think that I have become "personal" in any way, for what or why? And I would not have thought of questioning your skills. 
But if you understand it in that way, it's wrong. Nevertheless, I would apologize.

But in order not to waste any "unnecessary" bandwidth, I say goodbye to this discussion - at least until I am addressed again.

**Comment by lakeroe** [comment]
To check my oscillator accuracy I wrote a small program running an interrupt service routine at a 1 Hz rate (using the integrated ARM systick timer).
Within the ISR I increment a counter and output its value via serial port.
This program was running now for 5 days and I compared the start and end value against an official time reference (https://uhr.ptb.de/).
In the end the "clock" was off by ~ +22 seconds. This results in an inaccuracy of about 50 ppm which should be well within the specification of the library. Maybe I'll try to improve that by adjusting the two load capacitors ...

Anyway, for my purpose the library works more than enough.

**Comment by udoklein** [comment]
@lakeroe The logs you sent do not imply that your oscillator is not long term accurate enough. They imply that it is not stable enough. If you do not own a frequency counter but a second (crystal based) Arduino or other microcontroller you could check this like so: 

https://blog.blinkenlight.net/experiments/measurements/crystal-deviations-2/

The idea is to take a video of this setup. If the measurements just drift more or less evenly then everything is fine. However if your oscillator is unstable you would notice erratic drifts. Unstability may be introduced by many factors, in particular temperature and vibrations.

**Comment by nameoftherose** [comment]
@udoklein how does the adjustment value (printed by the Swiss_Army_Debug_Helper in mode Da) relates to crystal accuracy/stability;
For a crystal with Frequency Tolerance ±10ppm and Frequency Stability Over Operating Temperature Range  ±10ppm (according to manufacturer), I get an adjustment 39 - 40ppm. Is that reasonable? What adjustment values have been observed?

**Comment by udoklein** [comment]
Because KST was always a little bit on the slow side when plotting I created an IPythone notebook to speed up the processing of the log files. [You can find it here.](https://gist.github.com/udoklein/5797a41a227ad6aa0e6d66e1646444cd)

**Comment by udoklein** [comment]
Here are the first 4 plots I generated from lakeroes logs:
![dcf_2018-07-01_arm_aggressive](https://user-images.githubusercontent.com/5356932/42733762-847d1634-8837-11e8-9195-2b03897b51cd.png)
![dcf_2018-07-02_arm_standard](https://user-images.githubusercontent.com/5356932/42733764-86cb4a6e-8837-11e8-9294-5b3bf77a93b8.png)
![dcf_2018-07-03_arm_conservative](https://user-images.githubusercontent.com/5356932/42733766-888cb586-8837-11e8-9e20-1ded4d1f0f6b.png)
![dcf_2018-07-04_arm_paranoid](https://user-images.githubusercontent.com/5356932/42733767-8a06477e-8837-11e8-9dfc-ac7eeb0089db.png)

**Comment by udoklein** [comment]
@nameoftherose what is acceptable and what not is determined by the datasheet of the crystal. Usually I observe adjustments in the range of +/- 50 ppm or less. Thus 40 ppm is OK. However the algorithm assumes that the crystal is stable. Stability is not checked or tuned by the algorithm. Thus this value gives no indication if it is stable or not.

If you signal quality is somewhat reasonable you can check stability also in mode Ds by visual inspection. If you scrool through the log file and the signal drifts at a more or less constant rate, then the crystal is stable. If the rate is not almost constant then the crystal is unstable.

**Comment by nameoftherose** [comment]
@udoklein thank you

**Comment by lakeroe** [comment]
@udoklein
Regarding the stability of my oscillator I might do some more testing as you suggested (but can't really promise when).

Regarding the plots you've created from my logs: How do you interpret them ? Is your library working as expected ?

**Comment by udoklein** [comment]
Looking at the plots the library works as expected. However I also see that I need more sophisticated analysis tools to gain deeper insights. This may take some time.

**Comment by lakeroe** [comment]
Due to my poor oscillator long term accuracy I did another test.
On my ARM STM32F103 Blue Pill board I changed the original (probably chinese) oscillator to [this one here](https://www.digikey.at/product-detail/de/txc-corporation/7A-8.000MAAE-T/887-1740-1-ND/3518790) and also replaced the original load capacitors to 15pF.

Then I repeated my test from further above. This time the clock was off by about 3 seconds after 86 hours which equals to about 10 ppm. This is an improvement by a factor of 5 and I hope this also leads to an improved and more stable short term accuracy.

Using this change I started further dcf77 receive tests with different antenna positions.

**1. conservative_minute_quality**

21.07.2018 09:55h   Start, antenna horizontal and rotated by 90 degrees towards Frankfurt (best position)
                  approximately 1/2h after clock_state==synced I rotated the antenna by 90 degrees (worst position)
22.07.2018 20:00h antenna vertical (2nd best position)
23.07.2018  7:00h antenna horizontal and pointing directly towards Frankfurt again (worst position)
23.07.2018 20:25h End

-> No matter how I orientate the antenna, the time is ALWAYS accurate !!!

LOG-File: [dcf_2018-07-21_ARM.zip](https://github.com/udoklein/dcf77/files/2224161/dcf_2018-07-21_ARM.zip)

**2. aggressive_minute_quality (old behavior)**

23.07.2018 20:48h Start, antenna horizontal and rotated by 90 degrees towards Frankfurt (best position)
                  approximately 1/2h after clock_state==synced I rotated the antenna by 90 degrees (worst position)
24.07.2018 07:00h clock drift +3s
24.07.2018 16:38h End, decoded date and time completely wrong

LOG-File: [dcf_2018-07-23_ARM.zip](https://github.com/udoklein/dcf77/files/2224175/dcf_2018-07-23_ARM.zip)

-> Time is still drifting and also decoded wrong

Conclusion:
I think this experiment shows the importance of an accurate and stable oscillator and there impact.
Furthermore you can clearly see the improved library works as expected.

**Comment by nameoftherose** [comment]
@lakeroe now that you have a known accuracy crystal, will you please tell me what the output of frequency analysis command `Da` is?
@udoklein is direct transition from `sync ↔ free` and vice versa valid?

**Comment by lakeroe** [comment]
According to another test I have to revise my opinion. Even with paranoid_minute_quality I get wrong date/time.

**1. conservative_minute_quality**

24.07.2018 20:20h Start, antenna horizontal and rotated by 90 degrees towards Frankfurt (best position)
                  approximately 1/2h after clock_state==synced I rotated the antenna by 90 degrees (worst position)
25.07.2018 07:05h clock drift +8s
25.07.2018 16:24h End, clock drift +9s

LOG: [dcf_2018-07-24_ARM.zip](https://github.com/udoklein/dcf77/files/2232702/dcf_2018-07-24_ARM.zip)

**2. paranoid_minute_quality**

25.07.2018 16:30h Start, antenna horizontal and rotated by 90 degrees towards Frankfurt (best position)
25.07.2018 17:08h clock_state==synced
25.07.2018 18:24h antenna rotated by 90 degrees (worst position)
25.07.2018 21:40h clock_state==synced, time accurate
26.07.2018 16:27h End, time and date completly wrong

LOG: [dcf_2018-07-25_ARM.zip](https://github.com/udoklein/dcf77/files/2232703/dcf_2018-07-25_ARM.zip)

@nameoftherose
I can't do any testing on a UNO anymore. The one I had was borrowed and I had to return it.
Here is the output of "Da" (after about 90 minutes with antenna in optimum position):

confirmed_precision ?? adjustment, deviation, elapsed
0.5000 ppm, @+ , 5.8125 ppm, -2 ticks, 17 min + 52999 ticks mod 60000
54/100~0/600

Does this answer your question ?

**Comment by nameoftherose** [comment]
@lakeroe yes thank you!

**Comment by udoklein** [comment]
@nameoftherose The state engine is implemented in `DCF77_Local_Clock::process_1_Hz_tick`. The intended state transitions are documented here: [https://blog.blinkenlight.net/experiments/dcf77/local-clock/](https://blog.blinkenlight.net/experiments/dcf77/local-clock/) and the state diagram is here: [https://blinkenlightblog.files.wordpress.com/2013/02/c13_dcf77_clock_states.png](https://blinkenlightblog.files.wordpress.com/2013/02/c13_dcf77_clock_states.png). In particular there should be no direct transition _synced_ --> _free_. 

However it is possible to transition from _synced_ to _useless_ by calling the setup method after the clock has synced.

**Comment by udoklein** [comment]
@lakeroe: I was on vacation so I did not follow this thread closely. What is your current state of affairs? Does the library work as expected in the good position or does it fail also in the good position?

**Comment by nameoftherose** [comment]
In file [dcf_2018-07-23_ARM.zip](https://github.com/udoklein/dcf77/issues/31#issuecomment-407435745) line 17406 a transition from `synced to free` is recorded.

**Comment by udoklein** [comment]
You refer to this:

```
Decoded time: 18-07-23 1 21:31:56 CEST ..
  18.07.23(1,1)21:32:55 CET 1,1 4 p(9000-8644:22) s(191-27:23) m(200-124:10) h(188-122:9) wd(114-82:5) D(190-154:5) M(158-118:5) Y(130-105:3) 72,24,22,24
Clock state: synced
Tick: 2

     2656, XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX---+--------2XXXXXXXXXXXXXXXXXXXX

Decoded time: 18-07-23 1 21:31:57 CEST ..
  ??.??.??(?,?)??:??:255 CET 1,1 0 p(9000-9000:0) s(0-0:0) m(0-0:0) h(0-0:0) wd(0-0:0) D(0-0:0) M(0-0:0) Y(0-0:0) 72,24,22,255
Clock state: free
Tick: 150
```

This is indeed interesting but admissible. Have a look at [https://github.com/udoklein/dcf77/blob/master/dcf77.h#L1570](https://github.com/udoklein/dcf77/blob/master/dcf77.h#L1570) 
```

            if (clock_state == Clock::synced || clock_state == Clock::locked) {
                // the important part is 150 < 200,
                // otherwise it will fall through to free immediately after changing to unlocked
                if (tick >= 1150) {
                    // The 1 Hz pulse was locked but now
                    // it is definitely out of phase.
                    unlocked_seconds = 1;

                    // 1 Hz tick missing for more than 1200ms
                    clock_state = Clock::unlocked;
                    Clock_Controller::phase_lost_event_handler();
                }
            }

            if (clock_state == Clock::unlocked || clock_state == Clock::free) {
                if (tick >= 1000) {
                    tick -= 1000;
```

That is the state engine will transition from synced to unlocked and then immediately transition from unlocked to free. This happens during one cycle. You can argue if this is a direct transition or not. The explanation is as follows: the phase quality gets very low and the algorithm can not determine anymore if it just lost the phase or if it was already out of phase.

The point is that tries dead reckoning of the phase in state "unlocked". If this is not possible then it has to default to free. What does this imply? In state "unlocked" it can leverage the knowledge of the approximate time to still reliably synthesize the target signal that making it slightly easier to reaquire the signal. If this is not the case then it will more or less "cold start".

It can be argued if there is a better way for the transition from synced to unlocked to decrease the likelyhood that it defaults to free. This is a very tricky question and in my opinion not really worth the effort for optimization. I could be done though.

**Comment by nameoftherose** [comment]
@udoklein Thank you for taking the time to explain in such detail. A subtle point I would say. I agree with you it is not worth the effort.

**Comment by nameoftherose** [comment]
@udoklein thank you.
Transition to `free` happens only in [line 1603 dcf.h](https://github.com/udoklein/dcf77/blob/b8a0dc847a2d2dc802b232c4a7b96224600933c9/dcf77.h#L1603), and can only happen <br>`if (unlocked_seconds > max_unlocked_seconds)`, which is not fulfilled in this case as `unlocked_seconds = 1;` in line 1576.
But it is a subtle point ...

**Comment by udoklein** [comment]
Yesterday was to hot in Germany. I should have spotted this on my own. You are of course right. I think this should be handled in a separate issue (#34).

**Comment by udoklein** [comment]
The issue is caused because the value of max_unlocked_seconds is not 3000 or 30000 but 0 after initialization of the library. This in turn is because the setup() of the local_clock class is not properly called.

**Comment by udoklein** [comment]
@nameoftherose Thanks a lot for noticing this one. I fixed it by now :)

**Comment by nameoftherose** [comment]
@udoklein , I should have also spotted that this was caused by the uninitialized `max_unlocked_seconds`, it is too hot in Crete too ...

**Comment by lakeroe** [comment]
@udoklein
Sorry for my late answer but I've also been on vacation.
Concerning my current state of affairs:
I've currently running a test for about 27 hours. The antenna was in the good position all time and I didn't touch the whole setup. In the beginning everything was fine.
After about 21 hours I noticed a clock drift of about 30 seconds.
After about 27 hours the clock was in sync again but the output of Da showed

confirmed_precision ?? adjustment, deviation, elapsed
0.5000 ppm, @- , 400.0000 ppm, 7829 ticks, 332 min + 54830 ticks mod 60000
0/100~0/600

before it showed

confirmed_precision ?? adjustment, deviation, elapsed
0.5000 ppm, @+ , 3.9375 ppm, 50 ticks, 20 min + 4051 ticks mod 60000
54/100~0/600

LOG-file: [dcf_2018-08-08_ARM.zip](https://github.com/udoklein/dcf77/files/2273897/dcf_2018-08-08_ARM.zip)

So I can't really be satisfied.
Do you still think this is a problem of my oscillator ?
Could you please also describe in a few words the difference between the displayed 0.5000 ppm and 3.9375 ppm ?

**Comment by udoklein** [comment]
1 ppm = one part per million = 1 / 1 000 000 = 1e-6
Here the ppm refers to the local oscillator frequency. E.g. if your (crystal) oscillator would be running at 16 MHz, then 1 ppm = 16 Hz.

What you describe is a very clear indicator of a large clock drift. In my opinion your local oscillator is unstable. If you scroll through your file then you can even see (looking as the "scope" output) that your clock is stable for a while and then starts to drift with an excessively high rate. This looks a lot like a way out of spec oscillator. Unless you can bring up any really convincing argument why your oscillator is good I would say the oscillator is at fault with a probably >95%.

If you do not have a frequency counter to check if the oscillator is good you could benchmark it with a second Arduino like this: [https://blog.blinkenlight.net/experiments/measurements/crystal-deviations/](https://blog.blinkenlight.net/experiments/measurements/crystal-deviations/) or like that [https://blog.blinkenlight.net/experiments/measurements/crystal-deviations-2/](https://blog.blinkenlight.net/experiments/measurements/crystal-deviations-2/).

In my opinion this issue is a hardware issue and there will be no software fix for it at all.

**Comment by lakeroe** [comment]
Because I've already tested an original Arduino Uno and a few different STM32F103C8T6 boards and they behaved all pretty much the same, I can hardly believe it's an oscillator problem.
So I checked again if I ported the library properly to the STM32 and found following issue:
want_high_phase_lock_resolution was set to false (10 ms sample resolution), but dcf77_process() was called every 1ms

Do you think this explains anything ?

Maybe you could also have a look at the changes I made to the library (they are marked by * lakeroe *) ...
[dcf77_lakeroe.zip](https://github.com/udoklein/dcf77/files/2278275/dcf77_lakeroe.zip)

If this all does not help, are you interested in finding the root cause of the problem ?
I could send you an STM32-board+programming adapter and a DCF77 receiver ...

**Comment by nameoftherose** [comment]
@lakeroe , do you have a commercial dcf clock? How does it perform relative to the board? Can you experiment with an officially supported board?
I am running the library on a atmega on breadboard with run of the mill crystal and capacitors for the last 8 months almost continuously. I am located 2.200km from the transmitter. I have observed just 2 cases of hour errors and the clock corrected within an hour or so. The error between clock and computer was never more than 200ms at the end of `free` periods. It reaches `synced state` every night for 6-10 hours.
The only thing special is that the board power supply (from usb) is passed through LC filters.

**Comment by lakeroe** [comment]
For now I will not do any further testing, but maybe I will check my local oscillator in the future.
The library works for me by just resetting it everytime a valid time is received.

Anyway, thanks for your comments and help so far ...
